Source/Core/Core/HW/DVDInterface.cpp

```diff
@@ -301,6 +329,49 @@ static void TransferComplete(u64 userdata, int cyclesLate)
 	}
 }
 
+static void FinishExecuteReadCommand(u64 userdata, int cyclesLate)
+{
+	DVDReadCommand command;
+	command.is_valid = false;
+
+	// Because a scheduled event can't contain an entire DVDReadCommand,
+	// and because pointers in scheduled events do not work with savestates,
+	// an ID is passed instead, which is used to lookup the command in a deque
+	for (auto i = read_commands_to_execute.begin(); i != read_commands_to_execute.end(); ++i)
+	{
+		if (i->id == userdata)
+		{
+			if (i != read_commands_to_execute.begin())
+			{
+				// This shouldn't happen, and it might not even be a
+				// problem in case it happens, but it's checked for safety
+				WARN_LOG(DVDINTERFACE, "Read command %u was not first in deque. "
+				         "Are commands being executed out of order?", userdata);
+			}
+
+			command = *i;
+			read_commands_to_execute.erase(i);
+			break;
+		}
+	}
+
+	if (!command.is_valid)
+	{
+		// This shouldn't happen, but it's checked for safety
+		ERROR_LOG(DVDINTERFACE, "No valid read command was found in deque. Read %u will fail", userdata);
+	}
+	else
+	{
+		// Here is the actual disc reading
+		if (!DVDRead(command.DVD_offset, command.output_address, command.length, command.decrypt))
+			PanicAlertT("Can't read from DVD_Plugin - DVD-Interface: Fatal Error");
+	}
+
+	// The final step is to notify the emulated software that the command has been executed
+	// There isn't a ScheduleEvent_Immediate, so we use ScheduleEvent_Threadsafe_Immediate
+	CoreTiming::ScheduleEvent_Threadsafe_Immediate(command.callback_event_type, command.callback_userdata);
```

Comment:
I thought pointers don't work with savestates? Or does that only apply when having a pointer in a CoreTiming event's userdata? Either way, the pointer/event we use here is the one that has been passed to ExecuteCommand earlier, and the final part of ExecuteCommand requires a CoreTiming event in case the command isn't a read command, so the ExecuteCommand argument can't be changed to a pointer.